File: android/app/src/main/kotlin/jp/cloxs/bittra/ble/BlePeripheral.kt
package jp.cloxs.bitra.ble

import android.annotation.SuppressLint
import android.bluetooth.*
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseData
import android.bluetooth.le.AdvertiseSettings
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.ParcelUuid
import java.security.SecureRandom
import java.util.UUID

@SuppressLint("MissingPermission")
class BlePeripheral(
    private val context: Context,
    private val log: (String, String) -> Unit,
    private val onDeliveryReceipt: (String, Int) -> Unit
) {

    companion object {
        private const val INITIAL_HIGH_PERFORMANCE_MS = 15_000L
    }

    private val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
    private val adapter: BluetoothAdapter? = bluetoothManager.adapter
    private val advertiser = adapter?.bluetoothLeAdvertiser

    private var gattServer: BluetoothGattServer? = null
    private var isAdvertising = false
    private var isGattServiceReady = false
    private var startAdvertisingWhenReady = false
    private var currentAdvertiseMode: Int? = null
    private var advertiseDowngradeRunnable: Runnable? = null

    private var myNoticeId: String = ""
    private var myTeaser: String = ""
    private var myBody: String = ""

    private val pendingNotifyPackets = mutableListOf<ByteArray>()
    private var isNotifying = false
    private var subscribedDevice: BluetoothDevice? = null

    private val pendingNoticeByDevice = mutableMapOf<String, String>()
    private val acknowledgedDevicesByNotice = mutableMapOf<String, MutableSet<String>>()

    private val mainHandler = Handler(Looper.getMainLooper())

    private val senderId: Long by lazy {
        val prefs = context.getSharedPreferences("bittra_ble", Context.MODE_PRIVATE)
        if (prefs.contains("sender_id")) {
            prefs.getLong("sender_id", 1L)
        } else {
            var generated = SecureRandom().nextInt().toLong() and 0xFFFFFFFFL
            if (generated == 0L) generated = 1L
            prefs.edit().putLong("sender_id", generated).apply()
            generated
        }
    }

    private val advertiseCallback = object : AdvertiseCallback() {
        override fun onStartSuccess(settingsInEffect: AdvertiseSettings?) {
            isAdvertising = true
            val mode = currentAdvertiseMode
            log("ADV", "startAdvertising success mode=${advertiseModeName(mode)}")
            if (mode == AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY) {
                scheduleAdvertiseDowngrade()
            }
        }

        override fun onStartFailure(errorCode: Int) {
            cancelAdvertiseDowngrade()
            log("ADV", "startAdvertising fail: $errorCode")
            isAdvertising = false
            currentAdvertiseMode = null
        }
    }

    private val gattServerCallback = object : BluetoothGattServerCallback() {
        override fun onServiceAdded(status: Int, service: BluetoothGattService) {
            if (service.uuid != GATTProfile.SERVICE_UUID) return
            isGattServiceReady = status == BluetoothGatt.GATT_SUCCESS
            log("PERIPH", "GATT service ready status=$status")
            if (isGattServiceReady && startAdvertisingWhenReady) {
                startAdvertisingWhenReady = false
                startAdvertising(initialBoost = true)
            }
        }

        override fun onConnectionStateChange(device: BluetoothDevice, status: Int, newState: Int) {
            log("PERIPH", "connectionStateChange ${device.address} status=$status newState=$newState")
            if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                pendingNoticeByDevice.remove(device.address)
            }
        }

        override fun onCharacteristicWriteRequest(
            device: BluetoothDevice,
            requestId: Int,
            characteristic: BluetoothGattCharacteristic,
            preparedWrite: Boolean,
            responseNeeded: Boolean,
            offset: Int,
            value: ByteArray
        ) {
            when (characteristic.uuid) {
                GATTProfile.REQ_CHAR_UUID -> {
                    if (!value.contentEquals(byteArrayOf(0x01))) {
                        log("GATT", "invalid REQ payload bytes=${value.size}")
                        if (responseNeeded) {
                            gattServer?.sendResponse(
                                device,
                                requestId,
                                BluetoothGatt.GATT_REQUEST_NOT_SUPPORTED,
                                offset,
                                null
                            )
                        }
                        return
                    }
                    check(myNoticeId.isNotBlank()) { "REQ received without active noticeId" }
                    pendingNoticeByDevice[device.address] = myNoticeId
                    log("GATT", "REQ received from central noticeId=$myNoticeId")

                    if (responseNeeded) {
                        gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, value)
                    }
                    sendBodyPreviewThenRest(device)
                }

                GATTProfile.ACK_CHAR_UUID -> {
                    if (!value.contentEquals(byteArrayOf(0x02))) {
                        log("GATT", "invalid ACK payload bytes=${value.size}")
                        if (responseNeeded) {
                            gattServer?.sendResponse(
                                device,
                                requestId,
                                BluetoothGatt.GATT_REQUEST_NOT_SUPPORTED,
                                offset,
                                null
                            )
                        }
                        return
                    }

                    val noticeId = pendingNoticeByDevice.remove(device.address)
                    if (noticeId == null) {
                        log("GATT", "ACK without pending body device=${device.address}")
                        if (responseNeeded) {
                            gattServer?.sendResponse(
                                device,
                                requestId,
                                BluetoothGatt.GATT_FAILURE,
                                offset,
                                null
                            )
                        }
                        return
                    }

                    val acknowledgedDevices = acknowledgedDevicesByNotice.getOrPut(noticeId) {
                        mutableSetOf()
                    }
                    val added = acknowledgedDevices.add(device.address)
                    if (responseNeeded) {
                        gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, value)
                    }
                    if (added) {
                        log("GATT", "delivery ACK noticeId=$noticeId count=${acknowledgedDevices.size}")
                        onDeliveryReceipt(noticeId, acknowledgedDevices.size)
                    }
                }

                else -> {
                    if (responseNeeded) {
                        gattServer?.sendResponse(
                            device,
                            requestId,
                            BluetoothGatt.GATT_REQUEST_NOT_SUPPORTED,
                            offset,
                            null
                        )
                    }
                }
            }
        }

        override fun onDescriptorWriteRequest(
            device: BluetoothDevice,
            requestId: Int,
            descriptor: BluetoothGattDescriptor,
            preparedWrite: Boolean,
            responseNeeded: Boolean,
            offset: Int,
            value: ByteArray
        ) {
            if (BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE.contentEquals(value)) {
                subscribedDevice = device
            }
            if (responseNeeded) {
                gattServer?.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, value)
            }
        }

        override fun onNotificationSent(device: BluetoothDevice, status: Int) {
            super.onNotificationSent(device, status)
            if (status == BluetoothGatt.GATT_SUCCESS) {
                isNotifying = false
                flushNotifyQueue(device)
            } else {
                log("GATT", "onNotificationSent error $status")
                isNotifying = false
            }
        }
    }

    fun setContent(noticeId: String, teaser: String, body: String) {
        require(noticeId.isNotBlank()) { "noticeId must not be blank" }
        myNoticeId = noticeId
        myTeaser = teaser
        myBody = body
    }

    fun start() {
        check(myNoticeId.isNotBlank()) { "Cannot start advertising without noticeId" }
        if (PayloadCodec.normalizeTeaser(myTeaser).isEmpty()) {
            stop()
            log("PERIPH", "skip empty teaser advertisement")
            return
        }
        if (adapter == null || !adapter.isEnabled) {
            log("PERIPH", "Bluetooth not poweredOn yet")
            return
        }

        cancelAdvertiseDowngrade()
        if (isAdvertising) {
            advertiser?.stopAdvertising(advertiseCallback)
            isAdvertising = false
            currentAdvertiseMode = null
        }

        setupGATTIfNeeded()
        if (isGattServiceReady) {
            startAdvertising(initialBoost = true)
        } else {
            startAdvertisingWhenReady = true
        }
    }

    fun stop() {
        cancelAdvertiseDowngrade()
        startAdvertisingWhenReady = false
        if (isAdvertising) {
            advertiser?.stopAdvertising(advertiseCallback)
            isAdvertising = false
            log("PERIPH", "stopAdvertising")
        }
        currentAdvertiseMode = null
        pendingNotifyPackets.clear()
        pendingNoticeByDevice.clear()
        isNotifying = false
        subscribedDevice = null
        gattServer?.clearServices()
        gattServer?.close()
        gattServer = null
        isGattServiceReady = false
    }

    private fun setupGATTIfNeeded() {
        if (gattServer != null) return

        gattServer = bluetoothManager.openGattServer(context, gattServerCallback)
        isGattServiceReady = false

        val service = BluetoothGattService(GATTProfile.SERVICE_UUID, BluetoothGattService.SERVICE_TYPE_PRIMARY)

        val reqChar = BluetoothGattCharacteristic(
            GATTProfile.REQ_CHAR_UUID,
            BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE or BluetoothGattCharacteristic.PROPERTY_WRITE,
            BluetoothGattCharacteristic.PERMISSION_WRITE
        )

        val chunkChar = BluetoothGattCharacteristic(
            GATTProfile.CHUNK_CHAR_UUID,
            BluetoothGattCharacteristic.PROPERTY_NOTIFY,
            BluetoothGattCharacteristic.PERMISSION_READ
        )

        val ackChar = BluetoothGattCharacteristic(
            GATTProfile.ACK_CHAR_UUID,
            BluetoothGattCharacteristic.PROPERTY_WRITE,
            BluetoothGattCharacteristic.PERMISSION_WRITE
        )

        val cccDescriptor = BluetoothGattDescriptor(
            UUID.fromString("00002902-0000-1000-8000-00805f9b34fb"),
            BluetoothGattDescriptor.PERMISSION_READ or BluetoothGattDescriptor.PERMISSION_WRITE
        )
        chunkChar.addDescriptor(cccDescriptor)

        service.addCharacteristic(reqChar)
        service.addCharacteristic(chunkChar)
        service.addCharacteristic(ackChar)

        gattServer?.addService(service)
        log("PERIPH", "GATT service added")
    }

    private fun startAdvertising(initialBoost: Boolean) {
        val advertiseMode = if (initialBoost) {
            AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY
        } else {
            AdvertiseSettings.ADVERTISE_MODE_BALANCED
        }
        startAdvertisingWithMode(advertiseMode)
    }

    private fun startAdvertisingWithMode(advertiseMode: Int) {
        if (advertiser == null) return

        cancelAdvertiseDowngrade()
        currentAdvertiseMode = advertiseMode

        val settings = AdvertiseSettings.Builder()
            .setAdvertiseMode(advertiseMode)
            .setConnectable(true)
            .setTimeout(0)
            .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_MEDIUM)
            .build()

        val payload = PayloadCodec.encodeAdvert(myTeaser, senderId)

        val advertiseData = AdvertiseData.Builder()
            .addServiceUuid(ParcelUuid(GATTProfile.SERVICE_UUID))
            .setIncludeDeviceName(false)
            .build()

        val scanResponse = AdvertiseData.Builder()
            .addManufacturerData(GATTProfile.MANUFACTURER_ID, payload)
            .setIncludeDeviceName(false)
            .build()

        advertiser.startAdvertising(settings, advertiseData, scanResponse, advertiseCallback)
        log(
            "ADV",
            "startAdvertising requested mode=${advertiseModeName(advertiseMode)} teaser=${PayloadCodec.normalizeTeaser(myTeaser)} bytes=${payload.size}"
        )
    }

    private fun scheduleAdvertiseDowngrade() {
        cancelAdvertiseDowngrade()
        val runnable = Runnable {
            advertiseDowngradeRunnable = null
            if (!isAdvertising || currentAdvertiseMode != AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY) {
                return@Runnable
            }

            advertiser?.stopAdvertising(advertiseCallback)
            isAdvertising = false
            currentAdvertiseMode = null
            log("ADV", "initial 15s boost complete; switching to BALANCED")
            startAdvertisingWithMode(AdvertiseSettings.ADVERTISE_MODE_BALANCED)
        }
        advertiseDowngradeRunnable = runnable
        mainHandler.postDelayed(runnable, INITIAL_HIGH_PERFORMANCE_MS)
    }

    private fun cancelAdvertiseDowngrade() {
        advertiseDowngradeRunnable?.let { mainHandler.removeCallbacks(it) }
        advertiseDowngradeRunnable = null
    }

    private fun advertiseModeName(advertiseMode: Int?): String = when (advertiseMode) {
        AdvertiseSettings.ADVERTISE_MODE_LOW_LATENCY -> "LOW_LATENCY"
        AdvertiseSettings.ADVERTISE_MODE_BALANCED -> "BALANCED"
        AdvertiseSettings.ADVERTISE_MODE_LOW_POWER -> "LOW_POWER"
        else -> "UNKNOWN"
    }

    private fun sendBodyPreviewThenRest(device: BluetoothDevice) {
        pendingNotifyPackets.clear()
        isNotifying = false

        val bytes = myBody.toByteArray(Charsets.UTF_8)
        val preview = if (bytes.size > GATTProfile.PREVIEW_BYTES) bytes.copyOfRange(0, GATTProfile.PREVIEW_BYTES) else bytes
        val rest = if (bytes.size > GATTProfile.PREVIEW_BYTES) bytes.copyOfRange(GATTProfile.PREVIEW_BYTES, bytes.size) else ByteArray(0)

        val chunkPayloadSize = 17
        var seq = 0

        fun enqueueChunks(data: ByteArray) {
            var offset = 0
            while (offset < data.size) {
                val end = Math.min(data.size, offset + chunkPayloadSize)
                val slice = data.copyOfRange(offset, end)
                offset = end

                val packet = ByteArray(3 + slice.size)
                packet[0] = (seq and 0xFF).toByte()
                packet[1] = ((seq shr 8) and 0xFF).toByte()
                packet[2] = 0

                System.arraycopy(slice, 0, packet, 3, slice.size)
                pendingNotifyPackets.add(packet)
                seq++
            }
        }

        enqueueChunks(preview)
        enqueueChunks(rest)

        if (pendingNotifyPackets.isEmpty()) {
            pendingNotifyPackets.add(byteArrayOf(0x00, 0x00, 0x01))
        }

        val lastIndex = pendingNotifyPackets.size - 1
        val lastPacket = pendingNotifyPackets[lastIndex]
        lastPacket[2] = 0x01
        pendingNotifyPackets[lastIndex] = lastPacket

        log("GATT", "enqueue notify packets=${pendingNotifyPackets.size} bodyBytes=${bytes.size}")
        flushNotifyQueue(device)
    }

    private fun flushNotifyQueue(device: BluetoothDevice) {
        if (isNotifying || pendingNotifyPackets.isEmpty()) return

        isNotifying = true
        val packet = pendingNotifyPackets.removeAt(0)

        val service = gattServer?.getService(GATTProfile.SERVICE_UUID)
        val char = service?.getCharacteristic(GATTProfile.CHUNK_CHAR_UUID)

        if (char != null) {
            char.value = packet
            val success = gattServer?.notifyCharacteristicChanged(device, char, false) ?: false
            if (!success) {
                log("GATT", "notifyCharacteristicChanged failed, buffer might be full")
                pendingNotifyPackets.add(0, packet)
                isNotifying = false
                mainHandler.postDelayed({ flushNotifyQueue(device) }, 50)
            }
        } else {
            isNotifying = false
        }
    }
}

File: ios/Runner/BLEPeripheral.swift
import Foundation
import CoreBluetooth

final class BLEPeripheral: NSObject, CBPeripheralManagerDelegate {

  private var pm: CBPeripheralManager!
  private let log: (String, String) -> Void
  private let onDeliveryReceipt: (String, Int) -> Void

  private var reqChar: CBMutableCharacteristic!
  private var chunkChar: CBMutableCharacteristic!
  private var ackChar: CBMutableCharacteristic!
  private var service: CBMutableService!

  private var myNoticeId: String = ""
  private var myTeaser: String = ""
  private var myBody: String = ""

  private var pendingNotifyPackets: [Data] = []
  private var isNotifying: Bool = false
  private var shouldStartWhenPoweredOn: Bool = false
  private var isGattServiceReady: Bool = false
  private var startAdvertisingWhenReady: Bool = false

  private var pendingNoticeByCentral: [UUID: String] = [:]
  private var acknowledgedCentralsByNotice: [String: Set<UUID>] = [:]

  private lazy var senderId: UInt16 = {
    let defaults = UserDefaults.standard
    let key = "bittra_ble_sender_id"
    if defaults.object(forKey: key) != nil {
      return UInt16(truncatingIfNeeded: defaults.integer(forKey: key))
    }
    let generated = UInt16.random(in: 1...UInt16.max)
    defaults.set(Int(generated), forKey: key)
    return generated
  }()

  init(
    log: @escaping (String, String) -> Void,
    onDeliveryReceipt: @escaping (String, Int) -> Void
  ) {
    self.log = log
    self.onDeliveryReceipt = onDeliveryReceipt
    super.init()
    self.pm = CBPeripheralManager(delegate: self, queue: nil)
  }

  func setContent(noticeId: String, teaser: String, body: String) {
    precondition(!noticeId.isEmpty, "noticeId must not be empty")
    myNoticeId = noticeId
    myTeaser = teaser
    myBody = body
  }

  func start() {
    precondition(!myNoticeId.isEmpty, "Cannot start advertising without noticeId")
    guard !PayloadCodec.normalizeTeaser(myTeaser).isEmpty else {
      stop()
      log("PERIPH", "skip empty teaser advertisement")
      return
    }
    guard pm.state == .poweredOn else {
      shouldStartWhenPoweredOn = true
      log("PERIPH", "Bluetooth not poweredOn yet")
      return
    }
    if pm.isAdvertising {
      pm.stopAdvertising()
    }
    setupGATTIfNeeded()
    if isGattServiceReady {
      startAdvertising()
    } else {
      startAdvertisingWhenReady = true
    }
  }

  func stop() {
    shouldStartWhenPoweredOn = false
    startAdvertisingWhenReady = false
    pm.stopAdvertising()
    pendingNotifyPackets.removeAll()
    pendingNoticeByCentral.removeAll()
    isNotifying = false
    pm.removeAllServices()
    service = nil
    reqChar = nil
    chunkChar = nil
    ackChar = nil
    isGattServiceReady = false
    log("PERIPH", "stopAdvertising")
  }

  private func setupGATTIfNeeded() {
    if service != nil { return }

    reqChar = CBMutableCharacteristic(
      type: GATTProfile.reqCharUUID,
      properties: [.writeWithoutResponse],
      value: nil,
      permissions: [.writeable]
    )

    chunkChar = CBMutableCharacteristic(
      type: GATTProfile.chunkCharUUID,
      properties: [.notify],
      value: nil,
      permissions: []
    )

    ackChar = CBMutableCharacteristic(
      type: GATTProfile.ackCharUUID,
      properties: [.write],
      value: nil,
      permissions: [.writeable]
    )

    service = CBMutableService(type: GATTProfile.serviceUUID, primary: true)
    service.characteristics = [reqChar, chunkChar, ackChar]

    pm.removeAllServices()
    isGattServiceReady = false
    pm.add(service)
    log("PERIPH", "GATT service added")
  }

  private func startAdvertising() {
    let localName = PayloadCodec.encodeLocalName(teaser: myTeaser, senderId: senderId)

    pm.startAdvertising([
      CBAdvertisementDataServiceUUIDsKey: [GATTProfile.serviceUUID],
      CBAdvertisementDataLocalNameKey: localName
    ])
    log(
      "ADV",
      "startAdvertising requested teaser=\(PayloadCodec.normalizeTeaser(myTeaser)) localNameBytes=\(localName.utf8.count)"
    )
  }

  func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
    log("PERIPH", "state=\(peripheral.state.rawValue)")
    if peripheral.state == .poweredOn && shouldStartWhenPoweredOn {
      log("PERIPH", "Bluetooth is now poweredOn, starting delayed broadcast")
      shouldStartWhenPoweredOn = false
      start()
    }
  }

  func peripheralManager(_ peripheral: CBPeripheralManager,
                         didAdd service: CBService,
                         error: Error?) {
    guard self.service != nil, service === self.service else { return }
    isGattServiceReady = error == nil
    log("PERIPH", "GATT service ready error=\(error?.localizedDescription ?? "-")")
    if isGattServiceReady && startAdvertisingWhenReady {
      startAdvertisingWhenReady = false
      startAdvertising()
    }
  }

  func peripheralManagerDidStartAdvertising(_ peripheral: CBPeripheralManager,
                                             error: Error?) {
    if let error = error {
      log("ADV", "startAdvertising failed error=\(error.localizedDescription)")
    } else {
      log("ADV", "startAdvertising success")
    }
  }

  func peripheralManager(_ peripheral: CBPeripheralManager, didReceiveWrite requests: [CBATTRequest]) {
    for request in requests {
      if request.characteristic.uuid == GATTProfile.reqCharUUID {
        guard request.value == Data([0x01]) else {
          log("GATT", "invalid REQ payload")
          continue
        }
        precondition(!myNoticeId.isEmpty, "REQ received without active noticeId")
        pendingNoticeByCentral[request.central.identifier] = myNoticeId
        log("GATT", "REQ received from central noticeId=\(myNoticeId)")
        sendBodyPreviewThenRest()
        continue
      }

      if request.characteristic.uuid == GATTProfile.ackCharUUID {
        guard request.value == Data([0x02]) else {
          log("GATT", "invalid ACK payload")
          pm.respond(to: request, withResult: .requestNotSupported)
          continue
        }
        guard let noticeId = pendingNoticeByCentral.removeValue(forKey: request.central.identifier) else {
          log("GATT", "ACK without pending body central=\(request.central.identifier)")
          pm.respond(to: request, withResult: .unlikelyError)
          continue
        }

        var acknowledged = acknowledgedCentralsByNotice[noticeId, default: []]
        let inserted = acknowledged.insert(request.central.identifier).inserted
        acknowledgedCentralsByNotice[noticeId] = acknowledged
        pm.respond(to: request, withResult: .success)

        if inserted {
          log("GATT", "delivery ACK noticeId=\(noticeId) count=\(acknowledged.count)")
          onDeliveryReceipt(noticeId, acknowledged.count)
        }
      }
    }
  }

  func peripheralManagerIsReady(toUpdateSubscribers peripheral: CBPeripheralManager) {
    flushNotifyQueue()
  }

  private func sendBodyPreviewThenRest() {
    pendingNotifyPackets.removeAll()
    isNotifying = false

    let bytes = Data(myBody.utf8)
    let preview = bytes.prefix(GATTProfile.previewBytes)
    let rest = bytes.dropFirst(GATTProfile.previewBytes)

    // Keep each notification within the default ATT payload budget.
    // The packet header uses 3 bytes (seq 2B + flags 1B), leaving 17 bytes
    // for body data when the characteristic value budget is 20 bytes.
    let chunkPayloadSize = 17
    var seq: UInt16 = 0

    func enqueueChunks(_ data: Data) {
      var offset = 0
      while offset < data.count {
        let end = min(data.count, offset + chunkPayloadSize)
        let slice = data.subdata(in: offset..<end)
        offset = end

        var packet = Data()
        packet.append(UInt8(seq & 0xFF))
        packet.append(UInt8((seq >> 8) & 0xFF))
        packet.append(UInt8(0))
        packet.append(slice)
        pendingNotifyPackets.append(packet)
        seq += 1
      }
    }

    enqueueChunks(Data(preview))
    enqueueChunks(Data(rest))

    if pendingNotifyPackets.isEmpty {
      pendingNotifyPackets.append(Data([0x00, 0x00, 0x01]))
    }

    let lastIndex = pendingNotifyPackets.count - 1
    var lastPacket = pendingNotifyPackets[lastIndex]
    lastPacket[2] = 0x01
    pendingNotifyPackets[lastIndex] = lastPacket

    log("GATT", "enqueue notify packets=\(pendingNotifyPackets.count) bodyBytes=\(bytes.count)")
    flushNotifyQueue()
  }

  private func flushNotifyQueue() {
    guard !isNotifying else { return }
    isNotifying = true

    while !pendingNotifyPackets.isEmpty {
      let packet = pendingNotifyPackets[0]
      let ok = pm.updateValue(packet, for: chunkChar, onSubscribedCentrals: nil)
      if !ok {
        log("GATT", "updateValue blocked; wait ready")
        isNotifying = false
        return
      }
      pendingNotifyPackets.removeFirst()
    }
    log("GATT", "notify complete")
    isNotifying = false
  }
}
